core/src/smartcontracts/isi/permissions/combinators.rs

```diff
@@ -180,398 +108,30 @@ impl IsAllowed<Instruction> for CheckNested {
             | Instruction::Fail(_)
             | Instruction::ExecuteTrigger(_) => self.validator.check(authority, instruction, wsv),
-            Instruction::If(if_box) => {
-                self.check(authority, &if_box.then, wsv)
-                    .and_then(|_| match &if_box.otherwise {
-                        Some(this_instruction) => self.check(authority, this_instruction, wsv),
-                        None => Ok(()),
-                    })
-            }
+            Instruction::If(if_box) => self
+                .check(authority, &if_box.then, wsv)
+                .least_permissive_with(|| match &if_box.otherwise {
+                    Some(otherwise) => self.check(authority, otherwise, wsv),
+                    None => ValidatorVerdict::Skip,
+                }),
             Instruction::Pair(pair_box) => self
                 .check(authority, &pair_box.left_instruction, wsv)
-                .and(self.check(authority, &pair_box.right_instruction, wsv)),
-            Instruction::Sequence(sequence_box) => sequence_box
-                .instructions
-                .iter()
-                .try_for_each(|this_instruction| self.check(authority, this_instruction, wsv)),
-        }
-    }
-}
-
-fn check_all_validators_have_the_same_type(validators: &[IsAllowedBoxed]) -> Result<()> {
-    let first_type = if let Some(first) = validators.first() {
-        first.validator_type()
-    } else {
-        return Ok(());
-    };
-
-    for validator in validators.iter().skip(1) {
-        let validator_type = validator.validator_type();
-        if validator_type != first_type {
-            return Err(ValidatorTypeMismatch {
-                expected: first_type,
-                actual: validator_type,
-            }
-            .into());
-        }
-    }
-
-    Ok(())
-}
-
-/// A container for multiple permissions validators. It will succeed if all validators succeed.
-#[derive(Debug)]
-pub struct AllShouldSucceed {
-    validators: Vec<IsAllowedBoxed>,
-}
-
-impl AllShouldSucceed {
-    /// Create new [`AllShouldSucceed`]
-    ///
-    /// # Errors
-    /// If provided validators have different types.
-    /// Type of the first element in the `validators` is considered exemplary
-    pub fn new(validators: Vec<IsAllowedBoxed>) -> Result<Self> {
-        check_all_validators_have_the_same_type(&validators)?;
-        Ok(Self { validators })
-    }
-
-    fn check_type(&self, validator_type: ValidatorType) -> Result<()> {
-        if let Ok(self_type) = self.validator_type() {
-            if self_type != validator_type {
-                return Err(ValidatorTypeMismatch {
-                    expected: validator_type,
-                    actual: self_type,
+                .least_permissive_with(|| self.check(authority, &pair_box.right_instruction, wsv)),
+            Instruction::Sequence(sequence_box) => {
+                if sequence_box.instructions.is_empty() {
+                    ValidatorVerdict::Skip
```

Comment:
Hm.... I think it makes more sense than others variants. Why should we allow empty sequence? And why should we deny it? We don't know what to do with it, so let's just make no decision and skip

Reply:
Ok, make sense. I just think that if we allow empty sequence we can eliminate branching.

Reply:
Let's keep this discussion unresolved to see other people opinions

Reply:
I'm more concerned with how other people will understand the behaviour. `allow` and `deny` are obvious. `skip` is neither. When _should_ you skip? When is the validator going to `skip`? What should you do if your query is `skip`ped? 

Once these questions are answered it makes sense to make the API more complex to reduce branching

Reply:
For an external user query or other operation can't be skipped, because on the highest level there is a judge which returns `Ok` or `Err`.
The reasons of skipping you can find in `ValidatorVerdict::Skip` documentation.

But it still may vary from one implementation to another. So when constructing judge we should be sure how does every validator inside of it work: when it skips, when denies and when allows